File: leetcode/101-200/134-gas-station/bench/gas_station.py
"""Benchmark harness for LeetCode #134 — Gas Station.

Mirrors gas_station.kara algorithm-for-algorithm. Committed as the correctness
oracle; not a measured lane.
"""

NP = 8
N = 200000
ITERS = 1200


def can_complete(gas, cost):
    n = len(gas)
    total = 0
    tank = 0
    start = 0
    for i in range(n):
        d = gas[i] - cost[i]
        total += d
        tank += d
        if tank < 0:
            start = i + 1
            tank = 0
    return start if total >= 0 else -1


def lcg(seed, n, cap):
    out = []
    x = seed
    for _ in range(n):
        x = (x * 1103515245 + 12345) % 2147483648
        out.append((x // 65536) % cap)
    return out


def main():
    gases = []
    costs = []
    for j in range(NP):
        gases.append(lcg(j + 1, N, 100))
        costs.append(lcg(j + 100, N, 90 if j % 2 == 0 else 110))

    sink = 0
    for it in range(ITERS):
        idx = (it * 3) % NP
        sink = (sink + can_complete(gases[idx], costs[idx])) % 1000000007
    print(sink)


if __name__ == "__main__":
    main()

File: leetcode/101-200/134-gas-station/bench/gas_station.rs
// Benchmark harness for LeetCode #134 — Gas Station.
// Mirrors gas_station.kara algorithm-for-algorithm.

fn can_complete(gas: &[i64], cost: &[i64]) -> i64 {
    let n = gas.len();
    let mut total: i64 = 0;
    let mut tank: i64 = 0;
    let mut start: i64 = 0;
    for i in 0..n {
        let d = gas[i] - cost[i];
        total += d;
        tank += d;
        if tank < 0 {
            start = i as i64 + 1;
            tank = 0;
        }
    }
    if total >= 0 {
        start
    } else {
        -1
    }
}

fn lcg(seed: i64, n: i64, cap: i64) -> Vec<i64> {
    let mut out: Vec<i64> = Vec::new();
    let mut x = seed;
    for _ in 0..n {
        x = (x * 1103515245 + 12345) % 2147483648;
        out.push((x / 65536) % cap);
    }
    out
}

fn main() {
    let np: i64 = 8;
    let n: i64 = 200000;
    let iters: i64 = 1200;

    let mut gases: Vec<Vec<i64>> = Vec::new();
    let mut costs: Vec<Vec<i64>> = Vec::new();
    for j in 0..np {
        gases.push(lcg(j + 1, n, 100));
        if j % 2 == 0 {
            costs.push(lcg(j + 100, n, 90));
        } else {
            costs.push(lcg(j + 100, n, 110));
        }
    }

    let mut sink: i64 = 0;
    for it in 0..iters {
        let idx = ((it * 3) % np) as usize;
        sink = (sink + can_complete(&gases[idx], &costs[idx])) % 1000000007;
    }
    println!("{}", sink);
}
